### DISCUSSION TITLE
Timeout for entire SQE chain

### DISCUSSION THREAD

**Discussion by t3stme1x** [axboe/liburing#1101] [Q&A]
Greetings, io_uring community!

There is a read -> write SQE chain:
    prep_read(fd1, buf, buflen, -1) -> prep_write(fd2, buf, buf_size, -1) -> prep_write(fd3, buf, buf_size, -1) -> prep_write(fd4, buf, buf_size, -1) -> ...

We want to specify a timeout for the entire SQE chain. If the timeout is exceeded, the ring will cancel all pending operations in the chain.

The current workaround I have come up with is to issue a single prep_timeout() request along the SQE chain and race them together. If the timeout request completes first, all pending entries in the chain will be cancelled by prep_cancel() requests. Otherwise, the timeout request itself will be cancelled.

What would be the appropriate solution?

Do io_uring/liburing provide something like timeout_link that applies to the entire SQE chain, not just the previous SQE? If not, should we have it?

**Comment by cmazakas** [comment]
Ime, I've found linked timeouts to have a relatively heavy performance cost when used en masse.

What I do nowadays is start a multishot timer alongside my I/O operations that runs continuously. Then I add simple activity tracking to my I/O operations and when the timer periodically expires, it examines the time of the most recent activity and if required, will cancel the appropriate operation.

i.e. when an I/O op completes, I'll update the "last activity" with a call to `now()` or something like that which is basically free. If the last activity was too far in the past, I'll then manually start the cancellation process. This seems to work a'ight.

**Reply by isilence** [reply]
Yep, that's what I've been recommending all along, Per request (linked) timeouts are expensive, while it can be perfectly well implemented in userspace with some sort of timewheel just as @cmazakas mentioned. The kernel may try to optimise by ammortising with same time wheels, but that would be misguided, the user knows much better what the application pattern is.
The real question is how to have a helper so that such a time wheel (and a myriad of other cases) can be written once and reused by others. I guess I'll think about it

**Comment by YoSTEALTH** [comment]
edit: removed

**Reply by isilence** [reply]
That sounds just wrong. It'll wait for the timeout first and only then execute the rest of the link, that's not what the author was seeking for
